src/main/scala/com/fulcrumgenomics/bam/ErrorRateByReadPosition.scala

```diff
@@ -0,0 +1,170 @@
+  val FileExtension = ".error_rate_by_read_position.txt"
+}
+
+/** Gives the error rate for read one, read two, and across both reads, for the given position relative to the start of
+  * the read. For paired end reads, the read position is relative to the first base sequenced for the given end of the
+  * read.
+  *
+  * @param position the read position (1-based).
+  * @param error_rate_read_one the error rate of the first read.
+  * @param error_rate_read_two the error rate of the second read.
+  * @param error_rate_total the error rate across all reads at this read position.
+  * @param bases_read_one the total number of bases observed at this position for read one.
+  * @param bases_read_two the total number of bases observed at this position for read two.
+  * @param bases_total the total number of bases observed at this position across all reads.
+  */
+case class ErrorRateByReadPositionMetric
+( position: Int,
+  error_rate_read_one: Double,
+  error_rate_read_two: Double,
+  error_rate_total: Double,
+  bases_read_one: Long,
+  bases_read_two: Long,
+  bases_total: Long
+) extends Metric
+
+private class ErrorRateByReadPositionCounter {
+  val r1TotalBasesByReadPosition = new NumericCounter[Int]()
+  val r1ErrorBasesByReadPosition = new NumericCounter[Int]()
+  val r2TotalBasesByReadPosition = new NumericCounter[Int]()
+  val r2ErrorBasesByReadPosition = new NumericCounter[Int]()
+
+  def accept(rec: SAMRecord): Unit = {
+    val first: Boolean    = !rec.getReadPairedFlag || rec.getFirstOfPairFlag
+    val totalBasesByReadPosition = if (first) r1TotalBasesByReadPosition else r2TotalBasesByReadPosition
+    val errorBasesByReadPosition = if (first) r1ErrorBasesByReadPosition else r2ErrorBasesByReadPosition
+    val read              = maybeReverseComplement(rec.getReadBases.clone(), rc=rec.getReadNegativeStrandFlag)
+    val reference         = maybeReverseComplement(SequenceUtil.makeReferenceFromAlignment(rec, false), rc=rec.getReadNegativeStrandFlag)
+
+    var readPosition = 0
+    while (readPosition < read.length) {
+      val refBase  = reference(readPosition)
+      val readBase = read(readPosition)
+      // Make sure we are not at an insertion and not soft-clipped.  Also the bases must be valid
+      if ('-' != refBase && '0' != refBase && validBase(refBase) && validBase(readBase)) {
+        if (!SequenceUtil.basesEqual(read(readPosition), reference(readPosition))) { // mismatch
+          errorBasesByReadPosition.count(readPosition)
+        }
+        totalBasesByReadPosition.count(readPosition)
+      }
+      readPosition += 1
+    }
+  }
+
+  def metrics: Seq[ErrorRateByReadPositionMetric] = {
+    if (r1TotalBasesByReadPosition.isEmpty && r2TotalBasesByReadPosition.isEmpty) return Seq.empty
+    // create the metrics
+    val maxReadPosition = (r1TotalBasesByReadPosition.map(_._1) ++ r2TotalBasesByReadPosition.map(_._1)).max
+    Seq.range(0, maxReadPosition+1).map { readPosition =>
```

Comment:
I would change this to either `Range.inclusive(0, maxReadPosition)` or `(0 to maxReadPosition)`